### ISSUE TITLE
Regent: Affine helper breaks in PUC Lua

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#834]
We are apparently failing at this line, probably because the function being called here tries to call a Terra function:

https://github.com/StanfordLegion/legion/blob/master/language/src/regent/affine_helper.t#L111

Unfortunately, I think we need to do this constant folding in pure Lua. We cannot rely on being able to call LuaJIT FFI or Terra functions.

We should really test PUC Lua in CI if we can, at least in a limited capacity. Currently we have not ability to catch this.

**Comment by elliottslaughter** [comment]
Sample backtrace from Mario:

```
src/terralib.lua:548: attempt to call local 'ffiwrapper' (a userdata value)
stack traceback:
	src/terralib.lua:548: in function <src/terralib.lua:546>
	.../mdirenzo/legion/language/src/regent/affine_helper.t:94: in function <.../mdirenzo/legion/language/src/regent/affine_helper.t:87>
	(tail call): in function <(tail call):-1>
	...nal/mdirenzo/legion/language/src/regent/type_check.t:815: in function <...nal/mdirenzo/legion/language/src/regent/type_check.t:626>
	(tail call): in function <(tail call):-1>
	(tail call): in function <(tail call):-1>
	...nal/mdirenzo/legion/language/src/regent/type_check.t:3209: in function <...nal/mdirenzo/legion/language/src/regent/type_check.t:3208>
	(tail call): in function <(tail call):-1>
	(tail call): in function <(tail call):-1>
	(tail call): in function <(tail call):-1>
	...
	(tail call): in function <(tail call):-1>
	(tail call): in function <(tail call):-1>
	src/terralist.lua:150: in function 'map'
	...nal/mdirenzo/legion/language/src/regent/type_check.t:3923: in function 'block'
	...nal/mdirenzo/legion/language/src/regent/type_check.t:4650: in function <...nal/mdirenzo/legion/language/src/regent/type_check.t:4583>
	(tail call): in function <(tail call):-1>
	(tail call): in function <(tail call):-1>
	...xternal/mdirenzo/legion/language/src/regent/passes.t:69: in function '?'
	prometeo_chem.rg:96: in function <prometeo_chem.rg:32>
	prometeo-desugared.rg:282: in main chunk
```

**Comment by mariodirenzo** [comment]
In case it is relevant, the compilers that are currently producing the issue are `gnu/8.4.0`. I'm using LLVM 3.8 on an HPC with Power9 processors.
The same error 
```
src/terralib.lua:548: attempt to call local 'ffiwrapper' (a userdata value)
stack traceback:
	src/terralib.lua:548: in function <src/terralib.lua:546>
```
is reproduced even at the terra level by running the test https://github.com/terralang/terra/blob/master/tests/array.t

**Comment by elliottslaughter** [comment]
I put a potential fix into master via commit 33ac9b876003f9d732d27e74801bc43d1f9e7f40. Please test and let me know if it works.

**Comment by elliottslaughter** [comment]
Part 2 of this fix is in 71b9584a03aa53c4077f8808391bb9fb796b5d10, but there's still something going wrong in @mariodirenzo's code.

**Comment by elliottslaughter** [comment]
The fix is confirmed, the other issue is unrelated.
